lib/galaxy/webapps/galaxy/api/history_contents.py

```diff
@@ -790,8 +790,14 @@ def contents_near(
             # sqlalchemy DateTime columns are not timezone aware, so parse `since` into timezone-aware
             # datetime and then convert to naive datetime object representing UTC,
             # assuming history.update_time represents UTC time.
-            since_date = since.astimezone(datetime.timezone.utc).replace(tzinfo=None)
-            if history.update_time <= since_date:
+            # this is the right format, but changes the date itself resulting in logical errors
+
+            # Although this creates a date with the right format for SQLAlchemy, it changes the date
+            # data resulting in logical errors during testing. I fed back the history.update_time
+            # into this conversion and it fails the inequality
+            # since_date = since.astimezone(datetime.timezone.utc).replace(tzinfo=None)
```

Comment:
```suggestion
            since = since if since.tzinfo is None else since.astimezone(datetime.timezone.utc).replace(tzinfo=None)
```
pydantic parses timezone info if it is given (which means `since.tzinfo is not None`), in which case we do need to reset the object into a non-timezone aware object. (That's not been documented apart from the docstring here-> https://github.com/samuelcolvin/pydantic/blob/master/pydantic/datetime_parse.py#L179).

If you don't do this and you pass in a (also valid) iso8601 string like '2021-08-26T08:05:17.485495+00:00' you'll get
```
history.update_time < since
Traceback (most recent call last):
  File "<string>", line 1, in <module>
TypeError: can't compare offset-naive and offset-aware datetimes
```

Reply:
Also the comments above should be updated, it should probably say
```
            # sqlalchemy DateTime columns are not timezone aware, but `since` may be a timezone aware
            # DateTime object if a timezone offset is provided (https://github.com/samuelcolvin/pydantic/blob/5ccbdcb5904f35834300b01432a665c75dc02296/pydantic/datetime_parse.py#L179).
            # If a timezone is provided (since.tzinfo is not None) we convert to UTC and remove tzinfo so that comparison with history.update_time is correct.
```

Reply:
I'm following your argument, but I think it might be more straightforward to always send in a timezone-aware date so that we can use the built-in date parser, then simply add a timezone to history.update_time before comparison.

Reply:
The library we use should support everything that is standard iso8601, i don't think there's a point in limiting other consumers of our API to include the timezone offset. Also every other comparison to time columns is using naive datetime objects.

Reply:
I think we're miscommunicating, which is not surprising on a conversation about date-conversions.

Check out what I put up just now. I've actually removed any external requirements that the date have or doesn't have timezones. The only thing that's different from your plan is I'm letting the native python parse parse whatever it gets...

```python
since_str = kwd.pop('since', None)
if since_str:
    since = dateutil.parser.isoparse(since_str)
else:
    since = None
```

Then inside the handler we compare two dates that both have timezones


```python
if since:
    since = since.replace(tzinfo=datetime.timezone.utc) if since.tzinfo is None else since
    history_update_time_tz = history.update_time.replace(tzinfo=datetime.timezone.utc)
    
    if history_update_time_tz <= since:
        trans.response.status = 204
        return
```

Reply:
The only diffrence between your strategy and mine (I think) is that this doesn't require manual string manipulation to handle every kind of ISO8601 date? Hope that was clear.

Reply:
Now the api can handle any kind of ISO date with or without timezones:
```python
# test parsing for other standard is08601 formats
sample_formats = ['2021-08-26T15:53:02+00:00', '2021-08-26T15:53:02Z', '20210826T155302Z', '2002-10-10T12:00:00−05:00']
for date_str in sample_formats:
    encoded_date = urllib.parse.quote_plus(date_str) # handles pluses, minuses
    history_contents = self._get(f"/api/histories/{history_id}/contents/near/100/100?since={encoded_date}")
    assert history_contents.status_code != 400
```

Reply:
Now that the test is written, I'm gonna put your suggestion in, see if it passes. But the strategy before was resulting in a data change to the incoming which resulted in this situation:

history.update_time <= converted(history.update_time) => False because your conversion changed the date data in the process

Reply:
Ahhh I stand corrected, your version does indeed pass now. There must have been another problem before.  I hate dates so much.